Try to make flow control consuming and processing crisper

LPardue opened the pull request:
Fixes #193.

This mainly wiggles the existing text around and tries to fully qualify transport vs application to help explain the ways an implementation might do things, which is independent of how an application protocol might be designed (for example H3 doesn't require incremental processing but most people come to a realizatopm that it is the best way to do things).

martinthomson commented on draft-ietf-quic-applicability.md:
```diff
@@ -418,28 +418,34 @@ whether they become a problem depends on how implementations consume data and
-length-prefixed message format makes it easier for the data consumer to leave
-data unread in the receiver's buffers and thereby withhold flow control credit.
-If flow control limits prevent the remainder of a message from being sent, a
+Applications that use QUIC often have a data consumer that reads data from
+transport buffers. Timely updates of transport flow control credit can improve
+performance. A common flow control implementation technique is for a QUIC
```
Reorder these first two sentences:

```suggestion
Timely updates of transport flow control credit can improve performance.
Applications that use QUIC often have a data consumer that reads data from
transport buffers. A common flow control implementation technique is for a QUIC
```

martinthomson commented on draft-ietf-quic-applicability.md:
```diff
@@ -418,28 +418,34 @@ whether they become a problem depends on how implementations consume data and
+receiver to extend credit to the sender as the data consumer reads data.  Some
+implementations might have independent transport-layer and application-layer
+receive buffers. Consuming data does not always imply it is immediately
+processed. However, when data is consumed,
```
Incomplete sentence.

gorryfair commented:
The new text is good and clear, but it doesn't really explain that the when credit is released at the receiver, could we say a little more: it takes 1/2 RTT for cerdit to reach the sender, that causes new data to be released, for a large RTT, this means that credit has to be sufficient to avoid blocking or to be sent promptly, otherwise transfers will stall for periods of an RTT. (I know TCP does this stuff for the app and grows the window, but QUIC I suspect needs something to happen to release the credit). I know people closer to the Apps can write this sentence better than me; from my perspective I just see the effects when an app blocks or release credit in big chunks... over a large path RTT and the network performance suffers.

LPardue commented:
Thanks Gorry. I think I agree. How about I tweak "Timely updates can improve performance" paragraph to say "Timing of updates can affect performance". I can add a mention that the extension of credits on the data recever side is enacted by emitting MAX_DATA/MAX_STREAM_DATA frames, which take at least a 1/2 RTT to reach the data sender side.

I think the useful additional applicability statement to be made based on the above is that the data consumer read rate may be disjoint from the delivery rate and responsiveness of signalling channel back. For a highly asymmetric link, the download rate can be application-limited well below the theoretical bandwidth simply because an application is spending too much share of the time reading packets and not giving enough time to emit QUIC packets. 

Does that sound ok, or am I missing your ask?

gorryfair commented:
This is definitely heading in the direction I was hoping, it helps avoid people developing on a LAN and then discovering the effects on other paths later. Please add something.

mirjak reviewed (changes requested):
This PR is fine, only that there is an incomplete sentence...

LPardue commented:
I've reworked the paragraph with my attempt to address @gorryfair's comment. I've introduced the terms uplink and downlink, which I'm not totally happy with. If someone has a better suggestion I'm all ears.

mirjak commented on draft-ietf-quic-applicability.md:
```diff
@@ -418,28 +418,36 @@ whether they become a problem depends on how implementations consume data and
+always imply it is immediately processed. However, a common flow control
+implementation technique is to extend credit to the sender, by emitting MAX_DATA
+and/or MAX_STREAM_DATA frames, as data is consumed. Delivery of these frames
+is affected by uplink latency. If credit is not extended in a timely manner, the
```
```suggestion
is affected by the latency of the back channel from the receiver to the data sender.
If credit is not extended in a timely manner, the
```
Is this any better?

britram commented on draft-ietf-quic-applicability.md:
```diff
@@ -418,28 +418,37 @@ whether they become a problem depends on how implementations consume data and
+and/or MAX_STREAM_DATA frames, as data is consumed. Delivery of these frames
+is affected by the latency of the back channel from the receiver to the data sender.
+If credit is not extended in a timely manner, the
+sending application can be blocked, effectively throttling the downlink.
```
```suggestion
sending application can be blocked, effectively throttling the connection.
```
